### DISCUSSION TITLE
Question about k3s agent load balancer

### DISCUSSION THREAD

**Discussion by aisensiy** [k3s-io/k3s#4488] [Q&A] [answered]
There are articles talking about the embedded etcd HA solution of k3s like [this](https://blog.alexellis.io/bare-metal-kubernetes-with-k3s/). One of the key behavior is adding a load balancer solution (EIP like this article or LB from the clound provider) between the agents and masters:

```
k3s agent --> load balancer --> master
```

And the [architecture of k3s](https://rancher.com/docs/k3s/latest/en/architecture/#fixed-registration-address-for-agent-nodes) also show that a **Fixed Registration Address** is necessary.

While, after some research I found that k3s (at least v1.21.5+k3s2) have a internal agent load balancer (config at `/var/lib/rancher/k3s/agent/etc/k3s-agent-load-balancer.yaml`) which will auto update the master k8s api server list in it. So the out side load balancer is unnecessary?

And I also do a experiment like this:

1. Create a k3s cluster with 3 master nodes (master0 master1 master2) and 1 worker (worker0), and the worker agent joined the cluster from the master0 node.
2. Create a ssh session (session0) to worker0 with command `journal -u k3s-agent -f` to watch the log going.
3. Create a ssh session (session1) to master0 and run command `systemctl stop k3s`, then session0 (from the log tail) will show that the master0 is down and the agent load balancer is updated.
4. Create a ssh session (session2) to worker0 and run command `cat /var/lib/rancher/k3s/agent/etc/k3s-agent-load-balancer.yaml` will show there are only two records of master api endpoints over there.
5. Run command `systemctl start k3s` in session1 (the master0), the log in session0 will show that Agent load balancer updated and run command `cat /var/lib/rancher/k3s/agent/etc/k3s-agent-load-balancer.yaml` again in session2 there will show three ServerAddresses.

This behavior is so cool but I don't see any docs in k3s talking about this. So...do I misunderstand this behavior or there is little information about this behavior so that people still using a lb between agent and master?

**Comment by brandond** [comment, marked as answer]
> So the out side load balancer is unnecessary?

Our documentation lists a requirement for a "fixed registration endpoint" so that nodes do not rely on a single server being online in order to join the cluster. This endpoint could be a load-balancer or a DNS alias, it's up to you. This is only needed when nodes are registering to the cluster; once they have successfully joined, they use the client load-balancer to communicate directly with the servers without going through the registration endpoint.

**Reply by aisensiy** [reply]
This is a very clear and helpful reply. 👍
